src/components/application_manager/src/application_manager_impl.cc

```diff
@@ -808,6 +822,51 @@ void ApplicationManagerImpl::OnHMIStartedCooperation() {
   RefreshCloudAppInformation();
 }
 
+std::string ApplicationManagerImpl::PolicyIDByIconUrl(const std::string url) {
+  app_icon_map_lock_ptr_.Acquire();
+  for (auto& x : app_icon_map_) {
+    auto policy_id = x.first;
+    std::string icon_url = GetPolicyHandler().GetIconUrl(policy_id);
+    if (icon_url == url) {
+      LOG4CXX_DEBUG(logger_, "Matched icon url: " << url);
+    }
+    x.second.pending_request = false;
+    app_icon_map_lock_ptr_.Release();
+    return policy_id;
+  }
+  app_icon_map_lock_ptr_.Release();
+  return std::string("");
+}
+
+void ApplicationManagerImpl::SetIconExists(const std::string policy_id) {
+  app_icon_map_lock_ptr_.Acquire();
+  auto app_icon_it = app_icon_map_.find(policy_id);
+  if (app_icon_it != app_icon_map_.end()) {
+    LOG4CXX_DEBUG(logger_, "Updaing icon exists status");
+    app_icon_it->second.icon_exists = true;
```

Comment:
Part of me thinks that the entry should just be erased from the map if the icon is present (and never added to the map in the first place if it was already present on start)